presto-main/src/test/java/io/prestosql/operator/scalar/TestIpAddressFunction.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.operator.scalar;
+
+import org.testng.annotations.Test;
+
+import static io.prestosql.spi.type.BooleanType.BOOLEAN;
+
+public class TestIpAddressFunction
+        extends AbstractTestFunctions
+{
+    @Test
+    public void testIpAddressContains()
+    {
+        // Prefix length is 0
+        assertFunction("contains('10.0.0.1/0', IPADDRESS '0.0.0.0')", BOOLEAN, true);
+        assertFunction("contains('10.0.0.1/0', IPADDRESS '255.255.255.255')", BOOLEAN, true);
+
+        assertFunction("contains('::ffff:0a00:0001/0', IPADDRESS '0.0.0.0')", BOOLEAN, true);
+        assertFunction("contains('::ffff:0a00:0001/0', IPADDRESS '255.255.255.255')", BOOLEAN, true);
+
+        assertFunction("contains('10.0.0.1/0', IPADDRESS '::ffff:0000:0000')", BOOLEAN, true);
```

Comment:
I'm not sure these two cases should return `true`. A zero-length prefix is kind of non-sensical, so the result in these cases would likely be defined as a special case in the CIDR rules.

Reply:
Incidentally, looking at RFC 4632, it would seem a CIDR can't have an arbitrary ip address in the base part. All the bits beyond the prefix might need to be 0. I'll investigate this further.

BTW, here's a table with the possible CIDR notations for IPv4:

```
       notation       addrs/block      # blocks
       --------       -----------     ----------
       n.n.n.n/32               1     4294967296    "host route"
       n.n.n.x/31               2     2147483648    "p2p link"
       n.n.n.x/30               4     1073741824
       n.n.n.x/29               8      536870912
       n.n.n.x/28              16      268435456
       n.n.n.x/27              32      134217728
       n.n.n.x/26              64       67108864
       n.n.n.x/25             128       33554432
       n.n.n.0/24             256       16777216    legacy "Class C"
       n.n.x.0/23             512        8388608
       n.n.x.0/22            1024        4194304
       n.n.x.0/21            2048        2097152
       n.n.x.0/20            4096        1048576
       n.n.x.0/19            8192         524288
       n.n.x.0/18           16384         262144
       n.n.x.0/17           32768         131072
       n.n.0.0/16           65536          65536    legacy "Class B"
       n.x.0.0/15          131072          32768
       n.x.0.0/14          262144          16384
       n.x.0.0/13          524288           8192
       n.x.0.0/12         1048576           4096
       n.x.0.0/11         2097152           2048
       n.x.0.0/10         4194304           1024
       n.x.0.0/9          8388608            512
       n.0.0.0/8         16777216            256    legacy "Class A"
       x.0.0.0/7         33554432            128
       x.0.0.0/6         67108864             64
       x.0.0.0/5        134217728             32
       x.0.0.0/4        268435456             16
       x.0.0.0/3        536870912              8
       x.0.0.0/2       1073741824              4
       x.0.0.0/1       2147483648              2
       0.0.0.0/0       4294967296              1    "default route"
```

Reply:
From the RFC:

```
For example, the legacy "Class B" network 172.16.0.0, with an implied
   network mask of 255.255.0.0, is defined as the prefix 172.16.0.0/16,
   the "/16" indicating that the mask to extract the network portion of
   the prefix is a 32-bit value where the most significant 16 bits are
   ones and the least significant 16 bits are zeros.  Similarly, the
   legacy "Class C" network number 192.168.99.0 is defined as the prefix
   192.168.99.0/24; the most significant 24 bits are ones and the least
   significant 8 bits are zeros.
```